src/Umbraco.Web/PublishedContentQuery.cs

```diff
@@ -224,14 +225,51 @@ private IPublishedContent TypedDocumentById(int id, ContextualPublishedCache cac
         {
             var doc = cache.GetById(id);
             return doc;
-        }
+        }
+
+        private static readonly ConcurrentDictionary<Guid, int> _guidToIntLoopkup = new ConcurrentDictionary<Guid, int>();
 
         private IPublishedContent TypedDocumentById(Guid id, ContextualPublishedCache cache)
-        {
-            // todo: in v8, implement in a more efficient way
-            var legacyXml = UmbracoConfig.For.UmbracoSettings().Content.UseLegacyXmlSchema;
-            var xpath = legacyXml ? "//node [@key=$guid]" : "//* [@key=$guid]";
-            var doc = cache.GetSingleByXPath(xpath, new XPathVariable("guid", id.ToString()));
+        {
+            // Check if the loopkup is empty, if so populate it [LK]
+            if (_guidToIntLoopkup.Count == 0)
+            {
+                // TODO: Remove the debug profile logger
+                using (ApplicationContext.Current.ProfilingLogger.DebugDuration<PublishedContentQuery>("Populate GUID/INT lookup"))

```

Comment:
@zpqrtbnk @Shazwazza Do you see any issue with using `DebugDuration` here?
I'm _ever so slightly_ concerned about any performance bottlenecks here.